### PR TITLE
mds/tools: add set copyset available flag

### PR DESCRIPTION
**Pull request by bai-charisu**
Change-Id: I0b1e2e85f11790194f8ddb040b387d9b0e4ea0f4

### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: src/mds/topology/topology_item.h [hunk 1]
```diff
@@ -671,7 +671,8 @@ class CopySetInfo {
         epoch_(0),
         hasCandidate_(false),
         candidate_(UNINTIALIZE_ID),
-        dirty_(false) {}
+        dirty_(false),
+        available_(true) {}
```
Comment (ilixiaocui) [inline src/mds/topology/topology_item.h:675]: mds启动从etcd中load进来的时候，该copyset的available状态没处理？

File: src/tools/copyset_check_core.cpp [hunk 2]
```diff
@@ -55,9 +55,11 @@ int CopysetCheckCore::CheckOneCopyset(const PoolIdType& logicalPoolId,
     if (res != 0) {
         std::cout << "GetChunkServerListInCopySet from mds fail!"
                   << std::endl;
-        return -1;
+        return CheckResult::kOtherErr;
     }
-    bool isHealthy = true;
+    int majority = chunkserverLocation.size() / 2 + 1;
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/tools/copyset_check_core.cpp:60]: 不会，下面的CheckHealthOnLeader里会判断副本数量是否小于3

File: src/tools/copyset_check.cpp [hunk 3]
```diff
@@ -108,17 +108,19 @@ int CopysetCheck::RunCommand(const std::string& command) {
 }
 
 int CopysetCheck::CheckCopyset() {
-    int res = core_->CheckOneCopyset(FLAGS_logicalPoolId, FLAGS_copysetId);
-    if (res == 0) {
+    auto res = core_->CheckOneCopyset(FLAGS_logicalPoolId, FLAGS_copysetId);
+    int ret = 0;
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/tools/copyset_check.cpp:112]: 因为res不再是int类型了

File: src/mds/topology/topology.cpp [hunk 4]
```diff
@@ -1091,6 +1091,28 @@ int TopologyImpl::UpdateCopySetTopo(const CopySetInfo &data) {
...
 
+int TopologyImpl::SetCopySetAvalFlag(const CopySetKey &key, bool aval) {
+    ReadLockGuard rlockCopySetMap(copySetMutex_);
+    auto it = copySetMap_.find(key);
+    if (it != copySetMap_.end()) {
+        {
+            WriteLockGuard wlockCopySet(it->second.GetRWLockRef());
+            it->second.SetAvailableFlag(aval);
+        }
+        bool ret = storage_->UpdateCopySet(it->second);
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/topology/topology.cpp]: done

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 5]
```diff
@@ -212,19 +214,30 @@ bool AllocateChunkPolicy::AllocateChunkRandomInSingleLogicalPool(
...
     static std::random_device rd;   // generating seed for random number engine
     static std::mt19937 gen(rd());  // engin used: mersenne_twister_engine
     std::uniform_int_distribution<> dis(0, copySetIds.size() - 1);
 
-    for (uint32_t i = 0; i < chunkNumber; i++) {
+    for (uint32_t i = 0; i < chunkNumber;) {
         int randomCopySetIndex = dis(gen);
+        CopySetKey key(logicalPoolId, copySetIds[randomCopySetIndex]);
+        CopySetInfo csInfo;
+        if (!topology->GetCopySet(key, &csInfo)) {
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/topology/topology_chunk_allocator.cpp]: done

File: src/mds/topology/topology_service.cpp [hunk 6]
```diff
@@ -869,6 +869,70 @@ void TopologyServiceImpl::GetClusterInfo(
...
+void TopologyServiceImpl::ListUnAvailCopySets(
+                      google::protobuf::RpcController* cntl_base,
+                      const ListUnAvailCopySetsRequest* request,
+                      ListUnAvailCopySetsResponse* response,
+                      google::protobuf::Closure* done) {
+    brpc::ClosureGuard done_guard(done);
+    brpc::Controller* cntl =
+        static_cast<brpc::Controller*>(cntl_base);
+
+    LOG(INFO) << "Received request[log_id=" << cntl->log_id()
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/topology/topology_service.cpp:913]: request是空的，下面打印了response，response中有copyset list

File: src/tools/copyset_check_core.h [hunk 7]
```diff
@@ -123,7 +124,7 @@ class CopysetCheckCore {
     *
     * @return 健康返回0，不健康返回-1
     */
-    virtual int CheckOneCopyset(const PoolIdType& logicalPoolId,
+    virtual CheckResult CheckOneCopyset(const PoolIdType& logicalPoolId,
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/tools/copyset_check_core.h:127]: done

File: src/mds/topology/topology_item.h (hunk 1 above)
Comment (bai-charisu) [reply src/mds/topology/topology_item.h:675]: 已修改，加载和保存使用的是SerializeToString和ParseFromString

File: src/mds/topology/topology_item.cpp [hunk 8]
```diff
@@ -340,6 +341,7 @@ bool CopySetInfo::ParseFromString(const std::string &value) {
     logicalPoolId_ = data.logicalpoolid();
     copySetId_ = data.copysetid();
     epoch_ = data.epoch();
+    available_ = data.availflag();
```
Comment (ilixiaocui) [inline src/mds/topology/topology_item.cpp:344]: 升级时候老版本是没有这个字段的，需要处理一下

File: src/mds/topology/topology.h [hunk 9]
```diff
@@ -208,6 +210,8 @@ class Topology {
 
     virtual bool GetCopySet(CopySetKey key, CopySetInfo *out) const = 0;
 
+    virtual bool IsCopySetAvailable(CopySetKey key) const = 0;
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/topology/topology.h]: 去掉了

File: src/mds/topology/topology_chunk_allocator.cpp [hunk 10]
```diff
@@ -212,19 +214,26 @@ bool AllocateChunkPolicy::AllocateChunkRandomInSingleLogicalPool(
     std::vector<CopySetIdType> copySetIds,
     PoolIdType logicalPoolId,
     uint32_t chunkNumber,
-    std::vector<CopysetIdInfo> *infos) {
+    std::vector<CopysetIdInfo> *infos,
```
[the remark this answers is not in the document]
Comment (bai-charisu) [reply src/mds/topology/topology_chunk_allocator.cpp]: done

File: src/mds/topology/topology_item.cpp (hunk 8 above)
Comment (bai-charisu) [reply src/mds/topology/topology_item.cpp]: done

**Review by ilixiaocui — APPROVED** [review]

File: src/mds/topology/topology.cpp [hunk 11]
```diff
@@ -1091,6 +1091,30 @@ int TopologyImpl::UpdateCopySetTopo(const CopySetInfo &data) {
     }
 }
 
+int TopologyImpl::SetCopySetAvalFlag(const CopySetKey &key, bool aval) {
+    ReadLockGuard rlockCopySetMap(copySetMutex_);
+    auto it = copySetMap_.find(key);
+    if (it != copySetMap_.end()) {
+        auto copysetInfo = it->second;
```
Comment (xu-chaojie) [inline src/mds/topology/topology.cpp:1098]: 这里要先拿copyset的写锁，再storage_->UpdateCopySet，否则这里会把别人对copyset的修改覆盖掉

Comment (bai-charisu) [reply src/mds/topology/topology.cpp:1098]: done

File: src/mds/topology/topology.cpp [hunk 12]
```diff
@@ -1194,6 +1216,7 @@ void TopologyImpl::FlushCopySetToStorage() {
         }
     }
     for (auto &v : toUpdate) {
+        WriteLockGuard wlockCopySet(v.GetRWLockRef());
```
Comment (xu-chaojie) [inline src/mds/topology/topology.cpp:1219]: 这个地方改的不对，toUpdate里的copyset只是副本，这个锁起不了作用。

**Review by xu-chaojie — CHANGES_REQUESTED** [review]

**Comment by bai-charisu** [comment]
recheck

**Review by xu-chaojie — APPROVED** [review]

**Pull request merged**